source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOTypeaheadDynamicChoicesService.mm

```diff
@@ -0,0 +1,107 @@
+{
+    self = [super init];
+    if (self) {
+        _rootView = rootView;
+        _debouncer = [[ACOTypeaheadDebouncer alloc] initWithDelay:0.2];
+        _debouncer.delegate = self;
+        _global_queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
+        _inputElem = inputElem;
+        _typeaheadSearchDelegate = typeaheadSearchDelegate;
+    }
+
+    return self;
+}
+
+- (void)fetchChoicesFromDynamicSourceWithSearchQuery:(NSString *)searchQuery
+{
+    [_debouncer postInput:searchQuery];
+}
+
+- (void)dealloc
+{
+    _debouncer.delegate = nil;
+}
+
+- (NSDictionary *)getRequestPayloadForTypeaheadSearchWithQueryText:(NSString *)queryText
+                                                     withSkipValue:(NSNumber *)skipVal
+                                                       AndTopValue:(NSNumber *)topVal
+{
+    std::shared_ptr<BaseCardElement> elem = [_inputElem element];
+    std::shared_ptr<ChoiceSetInput> choiceSet = std::dynamic_pointer_cast<ChoiceSetInput>(elem);
+    std::shared_ptr<ChoicesData> choicesData = choiceSet->GetChoicesData();
+    NSString *choicesDataType = [NSString stringWithCString:AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::ChoicesDataType).c_str() encoding:NSUTF8StringEncoding];
+    NSString *datasetKey = [NSString stringWithCString:AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::Dataset).c_str() encoding:NSUTF8StringEncoding];
+    NSString *value = [NSString stringWithCString:AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::Value).c_str() encoding:NSUTF8StringEncoding];
+    NSString *dataset = [NSString stringWithCString:choicesData->GetDataset().c_str()
+                                 encoding:NSUTF8StringEncoding];
+    NSString *dataQuery = [NSString stringWithCString:choicesData->GetChoicesDataType().c_str()
+                                 encoding:NSUTF8StringEncoding];
+
+    NSMutableDictionary *requestPayload = [NSMutableDictionary new];
+    [requestPayload setDictionary: @{
+        choicesDataType   : dataQuery,
+        datasetKey        : dataset,
+        value             : queryText ?: @"",
+        @"skip"           : skipVal, //used for pagination.Usage - retrieve top n records by skipping first x records.
+        @"top"            : topVal
+    }];
+    return requestPayload;
+}
+
+#pragma mark - ACOTypeaheadDebouncerDelegate Methods specifically for dynamic typeahead
+
+- (void)debouncerDidSendOutput:(id)key
+{
+    if ([key isKindOfClass:NSString.class])
+    {
+        __weak typeof(self) weakSelf = self;
+        dispatch_async(_global_queue, ^{
+            __strong typeof(self) strongSelf = weakSelf;
+            if ([strongSelf->_rootView.acrActionDelegate respondsToSelector:@selector(onChoiceSetQueryChange:acoElem:completion:)]) {
+                NSDictionary *requestPayload = [strongSelf getRequestPayloadForTypeaheadSearchWithQueryText:key withSkipValue:[NSNumber numberWithInt:0] AndTopValue:[NSNumber numberWithInt:15]];
```

Comment:
If we are using a hardcoded values, why not just use const Number instances instead.

Reply:
replaced it with int.